Rework use_self impl based on ty::Ty comparison #3410 | Take 2

flip1995 opened the pull request:
This builds on top of #5531

I already reviewed and approved the commits by @montrivo. So only the review of my commits should be necessary. 

I would also appreciate your review @montrivo, since you are familiar with the challenges here.

Fixes #3410 and Fixes #4143 (same problem)
Fixes #2843
Fixes #3859 
Fixes #4734 and fixes #6221
Fixes #4305
Fixes #5078 (even at expression level now 🎉)
Fixes #3881 and Fixes #4887 (same problem)
Fixes #3909

Not yet: #4140 (test added)

All the credit for the fixes goes to @montrivo. I only refactored and copy and pasted his code.

changelog: rewrite [`use_self`] lint and fix multiple (8) FPs. One to go.

rust-highfive commented:
r? @ebroto

(rust_highfive has picked a reviewer for you, use r? to override)

montrivo commented on clippy_lints/src/use_self.rs:
```diff
@@ -53,10 +55,305 @@ declare_clippy_lint! {
+
+    fn check_body(&mut self, cx: &LateContext<'tcx>, body: &'tcx hir::Body<'_>) {
+        // `hir_ty_to_ty` cannot be called in `Body`s or it will panic (sometimes). But in bodies
+        // we can use `cx.typeck_results.node_type(..)` to get the `ty::Ty` from a `hir::Ty`.
```
Good to know! I wasn't aware of this.

flip1995 replied:
I discovered it randomly by searching for `hir_ty_to_ty` in rustc. This is used like that in `rustc_privacy`. https://github.com/rust-lang/rust/blob/8850893d96252172700727fe9e746c8e1ebd5853/compiler/rustc_privacy/src/lib.rs#L1246-L1258

ebroto commented:
Sorry I still have not reviewed this. It's on my radar but it's quite long and TBH it was sitting there for some time already :)

I'm giving priority to some PRs by new contributors and smaller ones, I expect I'll be able to get to this soonish.

flip1995 commented:
Don't worry, I'm also low on time and also prioritize reviewing over implementing things, if I got a bit of time.

bors commented:
:umbrella: The latest upstream changes (presumably #6338) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

ebroto commented:
I tried to rebase this but I'm not familiar enough with the changes.

flip1995 commented:
Oh this is still open 😄 I try to rebase it tomorrow.

flip1995 commented:
@ebroto I rebased it successfully. It was indeed quite a bit of work. The only remaining issue is #4140. I added it to the known problems section.

I think we should merge it like this, since I don't think I will get to it in the near future and it improves the situation of this lint immensely. Not sure if we should move the lint out of nursery now. WDYT?

ebroto reviewed (changes requested):
I just found some small stuff

---

The work done here is **extraordinary**, but besides #4140, #3881 #3909 #4887 are still relevant. Since we know the lint has still FPs, I would not move it from the nursery for now.

That said, the issues besides #4140 would be fixed by avoiding macros altogether and IMO we should go that way, as there's little benefit in linting local macros in this case. After that, just one known FP would remain.

Let me known what do you think.

flip1995 commented on clippy_lints/src/use_self.rs:
```diff
@@ -54,235 +56,395 @@ declare_clippy_lint! {
-
-            visitor.visit_ty(&impl_hir_ty);
+    fn check_ty(&mut self, cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>) {
+        if in_external_macro(cx.sess(), hir_ty.span)
```
For all the macro issues: This just has to be changed to `in_macro` (and the one below) and tests should be added. I'll do this in the next few days.

bors commented:
:umbrella: The latest upstream changes (presumably #6584) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

flip1995 commented:
I like that bors is reminding me about this every now and then.

ebroto commented:
r? @phansch (I'm leaving the team, so I'm reassigning my PRs to other active members)

phansch commented on clippy_lints/src/use_self.rs:
```diff
@@ -28,10 +29,11 @@ declare_clippy_lint! {
-    /// - False positive when using associated types (#2843)
-    /// - False positives in some situations when using generics (#3410)
+    /// - Unaddressed false negative in fn bodies of trait implementations
+    /// - False positive with with assotiated types in traits (#4140)
```
```suggestion
    /// - False positive with associated types in traits (#4140)
```

phansch commented on clippy_lints/src/use_self.rs:
```diff
@@ -54,235 +56,395 @@ declare_clippy_lint! {
+                // trait declaration. This is used to check if `Self` was used in the trait
+                // declaration.
+                //
+                // If `any`where in the `trait_sem_ty` the `self_ty` was used verbatim (as opposed
```
```suggestion
                // If anywhere in the `trait_sem_ty` the `self_ty` was used verbatim (as opposed
```

flip1995 replied:
That was intentional, because the `any` iterator method is used right below 😉 I can fix that though.

phansch reviewed (commented):
I can only agree with `ebroto` here: 'The work done here is extraordinary'. I just have some documentation nits. Let me know if you need any help with macro testing :slightly_smiling_face:

flip1995 commented:
Now #4140 is the only remaining issue.

phansch commented:
@bors r+ let's get this on the road :tada:

bors commented:
:pushpin: Commit e107f7618a5be1e9fb24d83892d486f2a2ff2535 has been approved by `phansch`

<!-- @bors r=phansch e107f7618a5be1e9fb24d83892d486f2a2ff2535 -->
<!-- homu: {"type":"Approved","sha":"e107f7618a5be1e9fb24d83892d486f2a2ff2535","approver":"phansch"} -->

bors commented:
:hourglass: Testing commit e107f7618a5be1e9fb24d83892d486f2a2ff2535 with merge a24fa494e211c8cafc1f40048c152089197b0c93...
<!-- homu: {"type":"BuildStarted","head_sha":"e107f7618a5be1e9fb24d83892d486f2a2ff2535","merge_sha":"a24fa494e211c8cafc1f40048c152089197b0c93"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1726274273)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/runs/1726274273","builder_name":"checks-action_test"} -->

flip1995 commented:
Oh no. I thought I fixed those with this PR. Well, now this PR will be stale a bit longer, if no one else takes over...

phansch commented:
@flip1995 I can try and take a look tomorrow, do you have any pointers where to start?

flip1995 commented on clippy_lints/src/use_self.rs:
```diff
@@ -54,235 +55,400 @@ declare_clippy_lint! {
-    fn nested_visit_map(&mut self) -> NestedVisitorMap<Self::Map> {
-        NestedVisitorMap::None
+    fn check_body(&mut self, cx: &LateContext<'tcx>, body: &'tcx hir::Body<'_>) {
+        // `hir_ty_to_ty` cannot be called in `Body`s or it will panic (sometimes). But in bodies
+        // we can use `cx.typeck_results.node_type(..)` to get the `ty::Ty` from a `hir::Ty`.
+        // However the `node_type()` method can *only* be called in bodies.
+        //
+        // This method implementation determines which types should get linted in a `Body` and
+        // which shouldn't, with a visitor. We could directly lint in the visitor, but then we
+        // could only allow this lint on item scope. And we would have to check if those types are
+        // already dealt with in `check_ty` anyway.
```
@phansch So I think, this is going wrong. Weirdly enough, I wrote, that `hir_ty_to_ty` cannot be used in bodies, but then go ahead and use in in`ty_from_hir_id`. Maybe it's a rebase gone wrong?

flip1995 commented on clippy_lints/src/use_self.rs:
```diff
@@ -54,235 +55,400 @@ declare_clippy_lint! {
+        // could only allow this lint on item scope. And we would have to check if those types are
+        // already dealt with in `check_ty` anyway.
+        if let Some(StackItem::Check { hir_id, .. }) = self.stack.last() {
+            let self_ty = ty_from_hir_id(cx, *hir_id);
```
If we're lucky, it is just 

```suggestion
            let self_ty = cx.typeck_results().node_type(*hir_id);
```

camsteffen commented on clippy_lints/src/use_self.rs:
```diff
@@ -54,235 +55,400 @@ declare_clippy_lint! {
+        // declaration. The collection of those types is all this method implementation does.
+        if_chain! {
+            if let ImplItemKind::Fn(FnSig { decl, .. }, ..) = impl_item.kind;
+            if let Some(StackItem::Check { impl_trait_ref_def_id: Some(def_id), .. }) = self.stack.last().copied();
```
Could you remove the `Option` from `impl_trait_ref_def_id` and just put `NoCheck` on the stack if it is `None`?

Also, I wonder if it would be better to avoid the stack altogether by using a separate Visitor which does not visit nested items. So, instead of `stack.push()`, you would have `MyItemVisitor().visit_item()` and other `check_` methods would be moved to `MyItemVisitor::visit_*`?

I wanted to bring up this idea because I also ran into a case where I need to have state-per-item and I was wondering which approach I should use. I hope that makes sense.

flip1995 replied:
> Could you remove the `Option` from `impl_trait_ref_def_id` and just put `NoCheck` on the stack if it is `None`?

Maybe? I'm not sure anymore, I implemented this ages ago. But I think you have to differentiate between trait impls and `NoCheck`s, because you have to prevent linting types in a trait impl, that are not defined as `Self` in the trait definition.

> Also, I wonder if it would be better to avoid the stack altogether by using a separate Visitor which does not visit nested items

The visitor method was used before. There are many other problems with that approach.  E.g. not being able to allow lints on expression level. Or that you still want to lint nested items and therefore still have to keep track of the latest state type -> stack. 

But the main problem I see with visitors is the redundant approach: You already are in a visitor context. Going into a visitor from there doesn't make much sense. (In general we should only use visitors to collect something, not to lint something)

camsteffen replied:
Thanks for explaining. I know you probably just want to get this PR out the door. 😃

flip1995 replied:
No problem. In fact, if you have more ideas or suggestions, please keep them coming. I'm petty much all out of ideas... :smile:

bors commented:
:umbrella: The latest upstream changes (presumably #6475) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

flip1995 commented on clippy_lints/src/use_self.rs:
```diff
@@ -54,235 +55,400 @@ declare_clippy_lint! {
-    #[must_use]
-    pub fn new(msrv: Option<RustcVersion>) -> Self {
-        Self { msrv }
+            if should_lint_ty(hir_ty, hir_ty_to_ty(cx.tcx, hir_ty), self_ty) {
```
So this `hir_ty_to_ty` ICEs. Which is weird, because IIRC we can't have any types inside a body at this point.

The span of the ICE suggests that it happens in closures, that need type inference.

camsteffen commented on clippy_lints/src/use_self.rs:
```diff
@@ -54,235 +55,400 @@ declare_clippy_lint! {
-        NestedVisitorMap::All(self.cx.tcx.hir())
+fn should_lint_ty(hir_ty: &hir::Ty<'_>, ty: Ty<'_>, self_ty: Ty<'_>) -> bool {
+    if_chain! {
+        if ty == self_ty;
```
Probably should be `TyS::same_type`?

giraffate commented:
ping from triage @phansch. I don't know my understanding is correct, but do you work on this?

camsteffen commented:
I think I have a working version in my branch at https://github.com/camsteffen/rust-clippy/tree/use-self. Basically just moved `types_to_lint` into the item stack because I think nested `impl Trait` or nested bodies were causing issues.

phansch commented:
> ping from triage @phansch. I don't know my understanding is correct, but do you work on this?

Oh sorry! Yes I wanted to build a minimal reproducer but didn't have the time yet. I'll hand it over to @camsteffen who's doing great work here :blue_heart:

flip1995 commented:
@bors try

bors commented:
:hourglass: Trying commit 52f98d832dc5ca982b9edad7807b2a9a9ddd13e0 with merge 3ad10ffec26f76e2cd7a786b0ed0c5b4fda52618...
<!-- homu: {"type":"TryBuildStarted","head_sha":"52f98d832dc5ca982b9edad7807b2a9a9ddd13e0","merge_sha":"3ad10ffec26f76e2cd7a786b0ed0c5b4fda52618"} -->

bors commented:
:sunny: Try build successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/1870689117), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/1870685715), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1870758887)
Build commit: 3ad10ffec26f76e2cd7a786b0ed0c5b4fda52618 (`3ad10ffec26f76e2cd7a786b0ed0c5b4fda52618`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/1870758887","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/1870689117","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/1870685715"},"merge_sha":"3ad10ffec26f76e2cd7a786b0ed0c5b4fda52618"} -->

flip1995 commented:
This passes now! Thanks @camsteffen! A second pair of eyes on this really helped. Did you manage to produce a minimal reproducer for the ICE?

camsteffen commented:
Glad I could help! No, sadly I could not reproduce the ICE without testing against `log` and `chalk`.

phansch commented:
LGTM! I think it would be nice to have a reproducer for the ICE, but then again, this fixes a lot of other issues with the lint. Let's get this in for now =)

@bors r+

bors commented:
:pushpin: Commit 52f98d832dc5ca982b9edad7807b2a9a9ddd13e0 has been approved by `phansch`

<!-- @bors r=phansch 52f98d832dc5ca982b9edad7807b2a9a9ddd13e0 -->
<!-- homu: {"type":"Approved","sha":"52f98d832dc5ca982b9edad7807b2a9a9ddd13e0","approver":"phansch"} -->

bors commented:
:hourglass: Testing commit 52f98d832dc5ca982b9edad7807b2a9a9ddd13e0 with merge 605e9ba3d7fa9e42afcdee5fd94c95f771693d07...
<!-- homu: {"type":"BuildStarted","head_sha":"52f98d832dc5ca982b9edad7807b2a9a9ddd13e0","merge_sha":"605e9ba3d7fa9e42afcdee5fd94c95f771693d07"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/1885146004), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/1885142118), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1885186812)
Approved by: phansch
Pushing 605e9ba3d7fa9e42afcdee5fd94c95f771693d07 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"phansch","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/1885186812","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/1885146004","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/1885142118"},"merge_sha":"605e9ba3d7fa9e42afcdee5fd94c95f771693d07"} -->

flip1995 commented:
I kind of expect that this still causes ICEs in some weird cases. We'll see with the next sync in 2 weeks...

tnielens commented:
Many thanks for the great work here! :heart:

Y-Nak commented:
I found minimal reproducer of the ICE on `Fix qpath_res call`: [7e1c1c1](https://github.com/rust-lang/rust-clippy/pull/6179/commits/7e1c1c154166d9eaa4184eab166b540e17ff26f3).
I think the ICE was caused by [here](https://github.com/flip1995/rust-clippy/blob/7e1c1c154166d9eaa4184eab166b540e17ff26f3/clippy_lints/src/use_self.rs#L202).

Minimal reproducer:
```rust
#![warn(clippy::use_self)]
#![allow(dead_code)]

fn main() {}

struct Foo {}

impl Foo {
    fn foo() -> Self {
        impl Foo {
            fn bar() {
            }
        }

        let x: _ = 1;

        Self {}
    }
}
```

camsteffen commented:
> I kind of expect that this still causes ICEs in some weird cases.

@flip1995 Yeah I'm not quite 100% confident either. In particular, the logic that prevents `hir_ty_to_ty` from being invoked in bodies is very indirect and fragile. It would be better to see logic like `if !self.in_body ..`. This would probably require using `check_body` and pushing onto a stack. Additionally this could probably replace some logic from `LintTyCollecter`.

> I found minimal reproducer of the ICE on `Fix qpath_res call`: [7e1c1c1](https://github.com/rust-lang/rust-clippy/pull/6179/commits/7e1c1c154166d9eaa4184eab166b540e17ff26f3).

@Y-Nak Nice! Want to open a PR with that?

> I think the ICE was caused by [here](https://github.com/flip1995/rust-clippy/blob/7e1c1c154166d9eaa4184eab166b540e17ff26f3/clippy_lints/src/use_self.rs#L202).

Yep that is also what I decided.

Y-Nak commented:
@camsteffen Yeah! I'll open a PR then.

flip1995 commented:
> In particular, the logic that prevents `hir_ty_to_ty` from being invoked in bodies is very indirect and fragile. It would be better to see logic like `if !self.in_body ..`.

Agreed. But I'm pretty sure I tried to just gate if we're in a body and it failed. I think the reason was, that arguments of closures/functions are in a body from the AST perspective, but not really in a body from the typeck perspective.

camsteffen commented:
Hmm. How about something like this?

* `check_item` => `self.in_body.push(false)`
* `check_fn` => `self.in_body.push(false)` (works for closures)
* `check_body` => `self.in_body.push(true)`
* `*_post` => `self.in_body.pop()`

camsteffen commented:
...or invert it and call it `in_header`. I think that is what we're actually interested in?

flip1995 commented:
I mean, we could try. I just know, that I didn't find a reliable way to differentiate between the part of the body we're currently in. Well, the `LintTyCollector` was my solution for this.
